#' ---
#' title: "Analysing GammaDelta cells in AD exps"
#' author: "gabrielet"
#' output: html_document
#' date: '`r format(Sys.Date(), "%d %B, %Y")`'
#' ---
#' 
#' ```{r setup, include=FALSE}
#' knitr::opts_chunk$set(
#'   tidy=TRUE,
#'   tidy.opts=list(width.cutoff=120),
#'   message=FALSE,
#'   warning=FALSE,
#'   eval=TRUE
#' )
#' ```

# Load libraries
library("bluster")
library("cluster")
library("compositions")
library("cowplot")
library("DropletUtils")
library("ggplot2")
library("gridExtra")
library("pheatmap")
library("plyr")
library("limma")
library("clusterProfiler")
library("org.Mm.eg.db")
library("RColorBrewer")
library("scater")
library("scran")
library("Seurat")
library("SingleCellExperiment")
library("openxlsx")
library("fgsea")
library("pathview")
library("ReactomePA")
library("DESeq2")
library("enrichplot")
library("DOSE")
library("HelpersMG")
library("gg.gap")
library("ddpcr")
library("data.table")
library("ggpubr")
library("edgeR")

# import plotting functions
source("../00_functions.R")

# initialise label
lbl <- 1

# set path
rawPath <- "/home/gabriele/work/cbmc/scrnaseq/raw_data_new/" ; expN <- "AD_batched"
saveIn <- paste0(rawPath, "batched_AD/")

# NOTE: we are using QCed data file which logcounts were normalised using sizeFactors
# but this data file contains also the information from the SLM clustering
# performed using the zinbwaved data, in order to have a good clustering but also all
# the genes, instead of the best 1000 used for zinbwave
perC <- 1000
kZinb <- 20	

# get normally clustered data
clusteredSce <- readRDS(paste0(saveIn, "QCed_data_with_clusters_", expN, "_K_", kZinb, "_top_", perC, ".Rds"))

# create fine_to_main labels as it was done for the zbWaved object in 03
clusteredSce$fine_to_main <- gsub(" ", "", unlist(lapply(strsplit(clusteredSce$pruned_fine, "\\("), `[[`, 1)))

##################### FULL DATASET

tab_alldata_labels <- table(clusteredSce$fine_to_main, clusteredSce$condition, clusteredSce$gender)

# get female full table ready to be exported
etabFemale <- as.data.frame.matrix(tab_alldata_labels[, , "F"])
eTabrnFemale <- cbind.data.frame(cellType=rownames(etabFemale), etabFemale)
eTabcnFemale <- rbind.data.frame(c("cluster", colnames(etabFemale)), eTabrnFemale)

# and male
etabMale <- as.data.frame.matrix(tab_alldata_labels[, , "M"])
eTabrnMale <- cbind.data.frame(cellType=rownames(etabMale), etabMale)
eTabcnMale <- rbind.data.frame(c("cluster", colnames(etabMale)), eTabrnMale)

# export the full table
write.xlsx(eTabcnFemale, paste0(saveIn, "female_cluster_content_full_labels_ALL_DATASET_exp_", expN, "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)
write.xlsx(eTabcnMale, paste0(saveIn, "male_cluster_content_full_labels_ALL_DATASET_exp_", expN, "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

# set cellTypes
cellType <- c("T.8|T.CD8", "Tgd", "Neutrophils", "\\(T.4|\\(T.CD4", "B cells", "DC|Macrophages|Microglia|Monocytes")
cellLab <-  c("CD8", "Tgd", "Neutrophils", "CD4", "B_cells", "Myeloid")

# loop through cell type
for (cT in seq(1, length(cellLab), by=1)){

	print("")
	print(paste0("analysing all the genes ", cellLab[cT], " for exp ", expN, " clustered with Zinbwave"))
	print("")
	
	# set paths
	saveUpDown <- paste0(saveIn, "enrichment_", cellLab[cT], "/UpDown/")
	saveImg <- paste0(saveIn, "figures/analysis_", cellLab[cT], "/")

	# check if saveIn directory exist, otherwise create it
	ifelse(dir.exists(saveIn), TRUE, dir.create(saveIn, recursive=T))
	ifelse(dir.exists(saveUpDown), TRUE, dir.create(saveUpDown, recursive=T))
	ifelse(dir.exists(saveImg), TRUE, dir.create(saveImg, recursive=T))

	# set palette with 12 colours for colouring plots
	colorsPal <- brewer.pal(12, "Paired")
	
	# plotting heatmap with cell type
	tab_fine_to_main <- table(Assigned=clusteredSce$fine_to_main, Cluster=clusteredSce$slm_0.5, Condition=clusteredSce$condition)

	# get 3x table ready to be exported
	eTabTX <- as.data.frame.matrix(tab_fine_to_main[, , "3xTG"])
	eTabRNTX <- cbind.data.frame(cellType=rownames(eTabTX), eTabTX)
	eTabCNTX <- rbind.data.frame(c("cluster", colnames(eTabTX)), eTabRNTX)

	# same for WT
	eTabWT <- as.data.frame.matrix(tab_fine_to_main[, , "WT"])
	eTabRNWT <- cbind.data.frame(cellType=rownames(eTabWT), eTabWT)
	eTabCNWT <- rbind.data.frame(c("cluster", colnames(eTabWT)), eTabRNWT)

	# export both tables
	write.xlsx(eTabCNTX, paste0(saveImg, "TX_cluster_content_ALL_CLUSTERS_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

	write.xlsx(eTabCNWT, paste0(saveImg, "WT_cluster_content_ALL_CLUSTERS_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

	##################### FULL

	tab_full_labels <- table(Assigned=clusteredSce$pruned_fine, Cluster=clusteredSce$slm_0.5, Condition=clusteredSce$condition)

	# get 3xTG full table ready to be exported
	etabFTX <- as.data.frame.matrix(tab_full_labels[, , "3xTG"])
	eTabrnFTX <- cbind.data.frame(cellType=rownames(etabFTX), etabFTX)
	eTabcnFTX <- rbind.data.frame(c("cluster", colnames(etabFTX)), eTabrnFTX)

	# and WT
	etabFWT <- as.data.frame.matrix(tab_full_labels[, , "WT"])
	eTabrnFWT <- cbind.data.frame(cellType=rownames(etabFWT), etabFWT)
	eTabcnFWT <- rbind.data.frame(c("cluster", colnames(etabFWT)), eTabrnFWT)

	# export the full table
	write.xlsx(eTabcnFTX, paste0(saveImg, "TX_cluster_content_full_labels_ALL_CLUSTERS_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

	write.xlsx(eTabcnFWT, paste0(saveImg, "WT_cluster_content_full_labels_ALL_CLUSTERS_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

	##################### FULL GENDER

	tab_GENDER_labels <- table(clusteredSce$pruned_fine, clusteredSce$condition, clusteredSce$gender)

	# get 3xTG full table ready to be exported
	etabFTX <- as.data.frame.matrix(tab_GENDER_labels[, , "M"])
	eTabrnFTX <- cbind.data.frame(cellType=rownames(etabFTX), etabFTX)
	eTabcnFTX <- rbind.data.frame(c("cluster", colnames(etabFTX)), eTabrnFTX)

	# and WT
	etabFWT <- as.data.frame.matrix(tab_GENDER_labels[, , "F"])
	eTabrnFWT <- cbind.data.frame(cellType=rownames(etabFWT), etabFWT)
	eTabcnFWT <- rbind.data.frame(c("cluster", colnames(etabFWT)), eTabrnFWT)

	# export the full table
	write.xlsx(eTabcnFTX, paste0(saveImg, "MALE_cluster_content_full_labels_ALL_CLUSTERS_BY_GENDER_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

	write.xlsx(eTabcnFWT, paste0(saveImg, "FEMALE_cluster_content_full_labels_ALL_CLUSTERS_BY_GENDER_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=F)

	########################################################################################################################################################################

	# get data set conditions, i.e. WT and 3xTG
	conD <- unique(clusteredSce$condition)

	# search for cellType
	cTPos <- grep(cellType[cT], clusteredSce$pruned_fine)

	print(paste0(cellLab[cT], " are found in ", length(unique(clusteredSce$slm_0.5[cTPos])), " different clusters"))

	fR <- table(clusteredSce$slm_0.5[cTPos])
	clustS <- names(fR[which(fR!=0)])

	################################################# SOME PLOTTING

	print(fR[which(fR!=0)])

	# Cluster composition plots using main labels

	df <- data.frame(table(clusteredSce$slm_0.5, clusteredSce$fine_to_main, clusteredSce$condition))

	colnames(df) <- c("Cluster", "CellType", "Condition", "Frequency")
	lims <- levels(factor(clusteredSce$fine_to_main))
	lims <- lims[order(table(clusteredSce$fine_to_main))]

	# plot!
	p <- plot_membership(df, cond=conD[2], col=colorsPal[2])
	# update label
	lbl <- lbl + 1
	png(paste0(saveImg, "04_0", lbl, "_", conD[2], "_cluster_composition_ALL_CLUSTERS_", expN, ".png"), type="cairo", units="in", width=8, height=6, pointsize=12, res=300)
		plot(p)
	dev.off()

	p <- plot_membership(df, cond=conD[1], col=colorsPal[6])
	# update label
	lbl <- lbl + 1
	png(paste0(saveImg, "04_0", lbl, "_", conD[1], "_cluster_composition_ALL_CLUSTERS_", expN, ".png"), type="cairo", units="in", width=8, height=6, pointsize=12, res=300)
		plot(p)
	dev.off()
	
	# Cluster composition plots SPLITTNG BY GENDER AND CONDITION

	conD <- unique(clusteredSce$conditionBYgender)

	dfReg <- data.frame(table(clusteredSce$slm_0.5, clusteredSce$fine_to_main, clusteredSce$conditionBYgender))
	colnames(dfReg) <- c("Cluster", "CellType", "Condition", "Frequency")
	
	write.xlsx(dfReg, paste0(saveIn, "composition_information_ALL_CLUSTERS_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)

	lims <- levels(factor(clusteredSce$fine_to_main))
	lims <- lims[order(table(clusteredSce$fine_to_main))]

	# plot!
	p <- plot_membership(dfReg, cond=conD[2], col=colorsPal[4])
	# update label
	lbl <- lbl + 1
	png(paste0(saveImg, "04_0", lbl, "_", conD[2], "_cluster_composition_ALL_CLUSTERS_", cellLab[cT], "_", expN, ".png"), type="cairo", units="in", width=8, height=6, pointsize=12, res=300)
		plot(p)
	dev.off()

	p <- plot_membership(dfReg, cond=conD[1], col=colorsPal[8])
	# update label
	lbl <- lbl + 1
	png(paste0(saveImg, "04_0", lbl, "_", conD[1], "_cluster_composition_ALL_CLUSTERS_", cellLab[cT], "_", expN, ".png"), type="cairo", units="in", width=8, height=6, pointsize=12, res=300)
		plot(p)
	dev.off()
	
	# plot!
	p <- plot_membership(dfReg, cond=conD[3], col=colorsPal[10])
	# update label
	lbl <- lbl + 1
	png(paste0(saveImg, "04_0", lbl, "_", conD[3], "_cluster_composition_ALL_CLUSTERS_", cellLab[cT], "_", expN, ".png"), type="cairo", units="in", width=8, height=6, pointsize=12, res=300)
		plot(p)
	dev.off()

	p <- plot_membership(dfReg, cond=conD[4], col=colorsPal[12])
	# update label
	lbl <- lbl + 1
	png(paste0(saveImg, "04_0", lbl, "_", conD[4], "_cluster_composition_ALL_CLUSTERS_", cellLab[cT], "_", expN, ".png"), type="cairo", units="in", width=8, height=6, pointsize=12, res=300)
		plot(p)
	dev.off()

	print(paste0("######## SEARCHING FOR ", cellType[cT], " IN THE WHOLE DATASET ########"))

	################################################# BACK TO THE EXPERIMENT

	# and subset the original single cell experiment using the name of the cellType cells.
	cTSubSet <- clusteredSce[, cTPos]
	
	# CONTE GREZZE ELEONORA
	raw_counts <- assay(cTSubSet, "counts")
	colnames(raw_counts) <- paste(colnames(cTSubSet), cTSubSet$condition, cTSubSet$gender, cTSubSet$pruned_fine, "cluster", cTSubSet$slm_0.5, sep="_")
	with_row_names <- cbind.data.frame(rownames(raw_counts), raw_counts)
	colnames(with_row_names) <- c("genes", colnames(raw_counts))
	
	write.table(with_row_names, paste0(saveIn, "counts_CONTE_GREZZE_CLUSTERING_GENERALE_exp_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
	#######################

	################################################# EXPORT THE NEW TABLE ELEONORA ASKED FEBRUARY #################################################

	superfull_table <- assay(cTSubSet, "logcounts")
	colnames(superfull_table) <- paste(colnames(cTSubSet), cTSubSet$condition, cTSubSet$gender, cellLab[cT], "cluster", cTSubSet$slm_0.5, sep="_")
	with_row_names <- cbind.data.frame(rownames(superfull_table), superfull_table)
	colnames(with_row_names) <- c("genes", colnames(superfull_table))
	
	write.table(with_row_names, paste0(saveIn, "logcounts_all_the_selected_cells_version_barbara_elena_eleonora_only_clustering_generale_PRIMA_DI_LOGNORMCOUNTS_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")

	# remove genes with too many zeros
	cTSubSet <- cTSubSet[rowMeans(assays(cTSubSet)$counts > 0) > 0.05,]

	# alternative to the filter above
	#	cTSubSet <- cTSubSet[rowSums(counts(cTSubSet) > 0) > 10,]

	# set minmean equal to 0.5 since this is the value used in the previous analysis, i.e. the one in the manuscript
	# compute sizeFactors
	cTSubSet <- computeSumFactors(cTSubSet, min.mean=0.5)
	cTSubSet <- logNormCounts(cTSubSet)
	
	################################################# EXPORT THE NEW TABLE ELENA ASKED #################################################

	superfull_table <- assay(cTSubSet, "logcounts")
	colnames(superfull_table) <- paste(colnames(cTSubSet), cTSubSet$condition, cTSubSet$gender, cTSubSet$pruned_fine, "cluster", cTSubSet$slm_0.5, sep="_")
	with_row_names <- cbind.data.frame(rownames(superfull_table), superfull_table)
	colnames(with_row_names) <- c("genes", colnames(superfull_table))
	
	write.table(with_row_names, paste0(saveIn, "logcounts_clustering_generale_version_barbara_elena_eleonora_only_ALL_CLUSTERS_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
	
	################################################# EXPORT THE NEW TABLE ELENA ASKED #################################################

	
	# exporting logcounts for Elena's violins
	subWT <- cTSubSet[, which(cTSubSet$condition=="WT")]
	subTX <- cTSubSet[, which(cTSubSet$condition=="3xTG")]

	logCsWT <- cbind.data.frame(genes=rownames(subWT), as.data.frame(logcounts(subWT)))
	logCsTX <- cbind.data.frame(genes=rownames(subTX), as.data.frame(logcounts(subTX)))
	
	write.table(logCsWT, paste0(saveIn, "logcounts_ALL_CLUSTERS_WT_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
	write.table(logCsTX, paste0(saveIn, "logcounts_ALL_CLUSTERS_3xTG_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")

#	write.xlsx(logCsWT, paste0(saveIn, "logcounts_ALL_CLUSTERS_WT_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)
#	write.xlsx(logCsTX, paste0(saveIn, "logcounts_ALL_CLUSTERS_3xTG_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)

	# exporting logcounts by condition for Elena's violins
	subWT <- cTSubSet[, which(cTSubSet$condition=="WT")]
	subTX <- cTSubSet[, which(cTSubSet$condition=="3xTG")]
	
	# MALE ONLY
	subWT_M <- subWT[, which(subWT$gender=="M")]
	subTX_M <- subTX[, which(subTX$gender=="M")]
	
	logCsWT_M <- cbind.data.frame(genes=rownames(subWT_M), as.data.frame(logcounts(subWT_M)))
	logCsTX_M <- cbind.data.frame(genes=rownames(subTX_M), as.data.frame(logcounts(subTX_M)))

	write.table(logCsWT_M, paste0(saveIn, "logcounts_ALL_CLUSTERS_WT_M_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
	write.table(logCsTX_M, paste0(saveIn, "logcounts_ALL_CLUSTERS_3xTG_M_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
	
#	write.xlsx(logCsWT_M, paste0(saveIn, "logcounts_ALL_CLUSTERS_WT_M_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)
#	write.xlsx(logCsTX_M, paste0(saveIn, "logcounts_ALL_CLUSTERS_3xTG_M_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)

	# F ONLY
	subWT_F <- subWT[, which(subWT$gender=="F")]
	subTX_F <- subTX[, which(subTX$gender=="F")]
	
	logCsWT_F <- cbind.data.frame(genes=rownames(subWT_F), as.data.frame(logcounts(subWT_F)))
	logCsTX_F <- cbind.data.frame(genes=rownames(subTX_F), as.data.frame(logcounts(subTX_F)))

	write.table(logCsWT_F, paste0(saveIn, "logcounts_ALL_CLUSTERS_WT_F_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
	write.table(logCsTX_F, paste0(saveIn, "logcounts_ALL_CLUSTERS_3xTG_F_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.csv"), row.names=F, quote=F, sep="\t")
#	write.xlsx(logCsWT_F, paste0(saveIn, "logcounts_ALL_CLUSTERS_WT_F_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)
#	write.xlsx(logCsTX_F, paste0(saveIn, "logcounts_ALL_CLUSTERS_3xTG_F_ONLY_", expN, "_", cellLab[cT], "_SLM_ZERO_CINQUE.xlsx"), overwrite=T, colNames=T)
	
	###############################################################################	###############################################################################	###############################################################################	###############################################################################	###############################################################################	###############################################################################	

	# COMPARING WT VERSUS 3XTG

	cTSce <- list()
	conD <- unique(clusteredSce$condition)
	# then get WT cellType
	cTSce[[conD[1]]] <- cTSubSet[, which(cTSubSet$condition==conD[1])]
	# and 3xTG
	cTSce[[conD[2]]] <- cTSubSet[, which(cTSubSet$condition==conD[2])]
	
	# set WT as a reference, to compute fold change
	cTSubSet$condition <- relevel(as.factor(cTSubSet$condition), ref="WT")

	if (ncol(cTSce[[conD[1]]]) >= 10 & ncol(cTSce[[conD[2]]]) >= 10) {
	
		print("finding markers")
	
		# FINDMARKERS for DEGS
		markersG <- findMarkers(cTSubSet, groups=cTSubSet$condition, pval.type="any", direction="any")

		cC <- "3xTG"
		chosen <- markersG[[cC]]

		# RELAXING THE THRESHOLD, is not done for meninges
		selfMade <- chosen[chosen$FDR < 0.05, ]
		#selfMade <- chosen[chosen$FDR < 0.1, ]

		# and export data
		finalGenes <- cbind.data.frame(gName=rownames(selfMade), selfMade)

		# sort by logfoldchange
		sortedFinal <- finalGenes[order(finalGenes$logFC, decreasing=T), ]

		# and export
#		write.table(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.csv"), col.names=T, row.names=F, sep="\t")
		write.xlsx(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.xlsx"), overwrite=TRUE)
	}
	
	# GABRIELA MALE VERSUS FEMALE
	
	cTSubSetGC <- cTSubSet
	
	cTSce <- list()
	conD <- unique(clusteredSce$gender)
	# then get WT cellType
	cTSce[[conD[1]]] <- cTSubSetGC[, which(cTSubSetGC$gender==conD[1])]
	# and 3xTG
	cTSce[[conD[2]]] <- cTSubSetGC[, which(cTSubSetGC$gender==conD[2])]
	
	# set FEMALE as a reference, to compute fold change
	cTSubSetGC$gender <- relevel(as.factor(cTSubSetGC$gender), ref="F")
	
	if (ncol(cTSce[[conD[1]]]) >= 10 & ncol(cTSce[[conD[2]]]) >= 10) {
	
		print("finding markers")
	
		# FINDMARKERS for DEGS
		markersG <- findMarkers(cTSubSetGC, groups=cTSubSetGC$gender, pval.type="any", direction="any")

		cC <- "M"
		chosen <- markersG[[cC]]

		# RELAXING THE THRESHOLD, is not done for meninges
		selfMade <- chosen[chosen$FDR < 0.05, ]
		#selfMade <- chosen[chosen$FDR < 0.1, ]

		# and export data
		finalGenes <- cbind.data.frame(gName=rownames(selfMade), selfMade)

		# sort by logfoldchange
		sortedFinal <- finalGenes[order(finalGenes$logFC, decreasing=T), ]

		# and export
#		write.table(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_MALE_VS_FEMALE_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.csv"), col.names=T, row.names=F, sep="\t")
		write.xlsx(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_MALE_VS_FEMALE_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.xlsx"), overwrite=TRUE)
	}	
	
	# GABRIELA MALE VERSUS FEMALE, 3xTG only
	
	# get 3xTG only
	cTSubSetGC_TX_only <- cTSubSet[, which(cTSubSet$condition=="3xTG")]
	
	# set WT as a reference, to compute fold change
	cTSubSetGC_TX_only$gender <- relevel(as.factor(cTSubSetGC_TX_only$gender), ref="F")
	
	cTSce <- list()
	conD <- unique(cTSubSetGC_TX_only$gender)

	if (ncol(cTSubSetGC_TX_only[, which(cTSubSetGC_TX_only$gender==conD[1])]) >= 10 & ncol(cTSubSetGC_TX_only[, which(cTSubSetGC_TX_only$gender==conD[2])]) >= 10) {
	
		print("finding markers")
	
		# FINDMARKERS for DEGS
		markersG <- findMarkers(cTSubSetGC_TX_only, groups=cTSubSetGC_TX_only$gender, pval.type="any", direction="any")

		cC <- "M"
		chosen <- markersG[[cC]]

		# RELAXING THE THRESHOLD, is not done for meninges
		selfMade <- chosen[chosen$FDR < 0.05, ]
		#selfMade <- chosen[chosen$FDR < 0.1, ]

		# and export data
		finalGenes <- cbind.data.frame(gName=rownames(selfMade), selfMade)

		# sort by logfoldchange
		sortedFinal <- finalGenes[order(finalGenes$logFC, decreasing=T), ]

		# and export
#		write.table(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_MALE_VS_FEMALE_3XTG_ONLY_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.csv"), col.names=T, row.names=F, sep="\t")
		write.xlsx(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_MALE_VS_FEMALE_3XTG_ONLY_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.xlsx"), overwrite=TRUE)
	}
	
	# GABRIELA MALE VERSUS FEMALE, WT only
	
	# get 3xTG only
	cTSubSetGC_WT_only <- cTSubSet[, which(cTSubSet$condition=="WT")]
	
	# set WT as a reference, to compute fold change
	cTSubSetGC_WT_only$gender <- relevel(as.factor(cTSubSetGC_WT_only$gender), ref="F")
	
	cTSce <- list()
	conD <- unique(cTSubSetGC_WT_only$gender)

	if (ncol(cTSubSetGC_WT_only[, which(cTSubSetGC_WT_only$gender==conD[1])]) >= 10 & ncol(cTSubSetGC_WT_only[, which(cTSubSetGC_WT_only$gender==conD[2])]) >= 10) {
	
		print("finding markers")
	
		# FINDMARKERS for DEGS
		markersG <- findMarkers(cTSubSetGC_WT_only, groups=cTSubSetGC_WT_only$gender, pval.type="any", direction="any")

		cC <- "M"
		chosen <- markersG[[cC]]

		# RELAXING THE THRESHOLD, is not done for meninges
		selfMade <- chosen[chosen$FDR < 0.05, ]
		#selfMade <- chosen[chosen$FDR < 0.1, ]

		# and export data
		finalGenes <- cbind.data.frame(gName=rownames(selfMade), selfMade)

		# sort by logfoldchange
		sortedFinal <- finalGenes[order(finalGenes$logFC, decreasing=T), ]

		# and export
#		write.table(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_MALE_VS_FEMALE_WT_ONLY_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.csv"), col.names=T, row.names=F, sep="\t")
		write.xlsx(sortedFinal, paste0(saveUpDown, "up_down_FindMarkers_ALL_CLUSTERS_MALE_VS_FEMALE_WT_ONLY_", cellLab[cT], "_CELLS_", expN, "_SLM_ZERO_CINQUE.xlsx"), overwrite=TRUE)
	}
}
